quesma/clickhouse/parser.go

```diff
@@ -4,41 +4,75 @@ package clickhouse
 
 import (
 	"fmt"
+	"quesma/logger"
 	"quesma/plugins"
+	"quesma/schema"
 	"quesma/util"
 	"slices"
 	"strings"
 )
 
 const NestedSeparator = "::"
 
+type CreateTableEntry struct {
+	ClickHouseColumnName string
+	ClickHouseType       string
+}
+
 // m: unmarshalled json from HTTP request
 // Returns nicely formatted string for CREATE TABLE command
-func FieldsMapToCreateTableString(namespace string, m SchemaMap, indentLvl int, config *ChTableConfig, nameFormatter plugins.TableColumNameFormatter) string {
-
+func FieldsMapToCreateTableString(m SchemaMap, config *ChTableConfig, nameFormatter plugins.TableColumNameFormatter, schemaMapping *schema.Schema) string {
 	var result strings.Builder
-	i := 0
-	for name, value := range m {
-		if namespace == "" {
-			result.WriteString("\n")
+
+	columnsFromJson := JsonToColumns("", m, 1, config, nameFormatter)
+	columnsFromSchema := SchemaToColumns(schemaMapping, nameFormatter)
+
+	first := true
+	for _, columnFromJson := range columnsFromJson {
+		if first {
+			first = false
+		} else {
+			result.WriteString(",\n")
+		}
+		result.WriteString(util.Indent(1))
+
+		if columnFromSchema, found := columnsFromSchema[schema.FieldName(columnFromJson.ClickHouseColumnName)]; found && !strings.Contains(columnFromJson.ClickHouseType, "Array") {
+			// Schema takes precedence over JSON (except for Arrays which are not currently handled)
+			result.WriteString(fmt.Sprintf("\"%s\" %s", columnFromSchema.ClickHouseColumnName, columnFromSchema.ClickHouseType))
+		} else {
+			result.WriteString(fmt.Sprintf("\"%s\" %s", columnFromJson.ClickHouseColumnName, columnFromJson.ClickHouseType))
 		}
 
+		delete(columnsFromSchema, schema.FieldName(columnFromJson.ClickHouseColumnName))
+	}
+
+	// There might be some columns from schema which were not present in the JSON
+	for _, column := range columnsFromSchema {
```

Comment:
We use `first` in both loops, should not be `first` reseted before this loop?

Reply:
The `first` variable refers here to being first line in `result` and both of the loops write to `result` (so a write in the second loop is no longer first if there was a write in the first loop)

Reply:
ok, thx for clarification